### ISSUE TITLE
Valgrind doesn't report allocations with jemalloc on nightly

### ISSUE THREAD

**Issue by apasel422** [rust-lang/rust#28224]
Compiling the following code with `opt-level=0`, Valgrind reports 26 allocs for `rustc 1.2.0 (082e47636 2015-08-03)`, but 5 allocs for `rustc 1.4.0-nightly (cd138dc44 2015-09-02)`:

``` rust
fn main() {
    let mut t = 0;

    for i in 0..10 {
        t += *Box::new(i);
    }

    assert_eq!(t, 45);
}
```

Perhaps this has something to do with #27400? In any case, it's harder to use Valgrind to debug Rust programs because of this.

**Comment by alexcrichton** [comment]
This is actually an optimization that LLVM is doing, it realizes that the allocation isn't necessary so it's elided entirely. As a result I believe this is working as intended, so closing.

**Comment by apasel422** [comment]
I'm not sure that's the case. Consider this program:
```rust
#![feature(test)]

extern crate test;

fn main() {
    for i in 0..100 {
        let foo = Box::new(i);
        test::black_box(foo);
    }
}
```
The assembly clearly contains calls to `exchange_malloc`, and running the code under `gdb` indicates that `exchange_malloc` is called 100 times:
```
(gdb) break _ZN4heap15exchange_malloc20he8623aa1084055ffJfaE
Breakpoint 1 at 0x4a90: file ../src/liballoc/heap.rs, line 123.
(gdb) ignore 1 1000
Will ignore next 1000 crossings of breakpoint 1.
(gdb) run
Starting program: /home/andrew/foo 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[Inferior 1 (process 1890) exited normally]
(gdb) info breakpoints
Num     Type           Disp Enb Address            What
1       breakpoint     keep y   0x0000555555558a90 in foo::heap::exchange_malloc at ../src/liballoc/heap.rs:123
	breakpoint already hit 100 times
	ignore next 900 hits
```

**Comment by apasel422** [comment]
But Valgrind reports:
```
==1899== Memcheck, a memory error detector
==1899== Copyright (C) 2002-2013, and GNU GPL'd, by Julian Seward et al.
==1899== Using Valgrind-3.10.1 and LibVEX; rerun with -h for copyright info
==1899== Command: ./foo
==1899== 
==1899== 
==1899== HEAP SUMMARY:
==1899==     in use at exit: 0 bytes in 0 blocks
==1899==   total heap usage: 5 allocs, 5 frees, 976 bytes allocated
==1899== 
==1899== All heap blocks were freed -- no leaks are possible
==1899== 
==1899== For counts of detected and suppressed errors, rerun with: -v
==1899== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

**Comment by alexcrichton** [comment]
Hm something fishy may indeed be going on here, it looks like valgrind is no longer tracking jemalloc? If you use `alloc_system` for now you'll get accurate tracking via valgrind, I'll reopen for more investigation into jemalloc though.

(also tweaking the title a bit)

**Comment by apasel422** [comment]
Using `alloc_system` indeed causes valgrind to report the correct information.

**Comment by shepmaster** [comment]
Sorry to jump in here, but how did you ever get valgrind to report jemalloc stats in the first place? On OS X, I've *never* successfully gotten any useful stats.

**Comment by apasel422** [comment]
I think the issue is that Valgrind can't track calls to jemalloc's `mallocx`-family of functions. There is a way to get Valgrind to track them, but it involves some custom headers: http://valgrind.org/docs/manual/mc-manual.html#mc-manual.mempools

**Comment by apasel422** [comment]
See also https://github.com/jemalloc/jemalloc/blob/master/include/jemalloc/internal/valgrind.h.

**Comment by apasel422** [comment]
jemalloc is apparently not being built with `--enable-valgrind` (which makes sense for performance reasons):

foo.c:
```C
#include <stdio.h>
#include <stdlib.h>

extern void je_mallctl(const char *, void *, size_t *, void *, size_t);

int main() {
    int valgrind = 6;
    size_t len = 1;
    je_mallctl("config.valgrind", &valgrind, &len, NULL, 0);
    printf("%d\n", valgrind);
}
```
```sh
> gcc foo.c -L/usr/local/lib/rustlib/x86_64-unknown-linux-gnu/lib/ -l:liballoc_jemalloc-35017696.rlib -lpthread
> ./a.out 
0
```

**Comment by apasel422** [comment]
Building nightly with
```sh
JEMALLOC_FLAGS='--enable-valgrind' make
```
results in the correct Valgrind behavior.

Not sure how we want to address this, but providing multiple versions of `liballoc_jemalloc` seems like a possibility.

**Comment by shepmaster** [comment]
FWIW, I ran your example program with `rustc 1.5.0-dev (5f06607e2 2015-09-29)` and `valgrind-3.11.0.SVN` on OS X, and it seems to track allocations:

```
==28285== HEAP SUMMARY:
==28285==     in use at exit: 35,221 bytes in 426 blocks
==28285==   total heap usage: 611 allocs, 185 frees, 42,405 bytes allocated
```

To my knowledge, I have not specified any custom flags when building. I also see that massif reports heap allocations over time, so I have some hope that things are just magically working together now!

**Comment by apasel422** [comment]
Valgrind on Ubuntu still indicates that this code, compiled with `opt-level=0` on nightly has only 5 allocations:
```rust
#![feature(alloc_jemalloc)]
#![feature(test)]

extern crate alloc_jemalloc;
extern crate test;

fn main() {
    for i in 0..100 {
        let foo = Box::new(i);
        test::black_box(foo);
    }
}
```
Removing the allocator has the same effect, while changing to `alloc_system` reports 112 allocations.

**Comment by Hansyperman** [comment]
I think I reported this same issue on reddit: rust 1.8 stable lost valgrind support
https://www.reddit.com/r/rust/comments/4fddre/i_lost_valgrind_in_rust_18/

I have a mixed rust/extern "C" program. I know it leaks memory and use valgrind to verify behaviour. When I upgraded to rust stable 1.8, it proudly claims not to leak anymore, and even stops allocating memory altogether! While I'm impressed by rusts memory safety guarantees, this is a bit too much for me ;-)

Anyone knows what happens and how to make it un-happen? I read somewhere they integrated jemalloc, so maybe that broke the monitoring allocator from valgrind?

An example on debian 8.4, x86-64:

    extern "C" {
      fn malloc(len: usize) -> *mut u8;
    }
    
    fn main() {
      println!("Allocated at {:?}",unsafe{malloc(100)});
    }

On rust 1.8:

    me@aruba:~/atest$ rustc --version
    rustc 1.8.0 (db2939409 2016-04-11)
    me@aruba:~/atest$ cargo build
    me@aruba:~/atest$ valgrind  target/debug/atest
    ==2921== Memcheck, a memory error detector
    ==2921== Copyright (C) 2002-2013, and GNU GPL'd, by Julian Seward et al.
    ==2921== Using Valgrind-3.10.0 and LibVEX; rerun with -h for copyright info
    ==2921== Command: target/debug/atest
    ==2921== 
    Allocated at 0x6035000
    ==2921== 
    ==2921== HEAP SUMMARY:
    ==2921==     in use at exit: 0 bytes in 0 blocks
    ==2921==   total heap usage: 0 allocs, 0 frees, 0 bytes allocated
    ==2921== 
    ==2921== All heap blocks were freed -- no leaks are possible
    ==2921== 
    ==2921== For counts of detected and suppressed errors, rerun with: -v
    ==2921== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)

On rust 1.7

    me@aruba:~/atest$ rustc --version
    rustc 1.7.0 (a5d1e7a59 2016-02-29)
    me@aruba:~/proggies/atest$ cargo build
    Compiling atest v0.1.0 (file:///home/hans/proggies/atest)
    me@aruba:~/atest$ valgrind  target/debug/atest
    ==5975== Memcheck, a memory error detector
    ==5975== Copyright (C) 2002-2013, and GNU GPL'd, by Julian Seward et al.
    ==5975== Using Valgrind-3.10.0 and LibVEX; rerun with -h for copyright info
    ==5975== Command: target/debug/atest
    ==5975== 
    Allocated at 0x5e00570
    ==5975== 
    ==5975== HEAP SUMMARY:
    ==5975==     in use at exit: 100 bytes in 1 blocks
    ==5975==   total heap usage: 7 allocs, 6 frees, 1,124 bytes allocated
    ==5975== 
    ==5975== LEAK SUMMARY:
    ==5975==    definitely lost: 100 bytes in 1 blocks
    ==5975==    indirectly lost: 0 bytes in 0 blocks
    ==5975==      possibly lost: 0 bytes in 0 blocks
    ==5975==    still reachable: 0 bytes in 0 blocks
    ==5975==         suppressed: 0 bytes in 0 blocks
    ==5975== Rerun with --leak-check=full to see details of leaked memory
    ==5975== 
    ==5975== For counts of detected and suppressed errors, rerun with: -v
    ==5975== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)

**Comment by alexcrichton** [comment]
The tools team discussed this issue during triage yesterday and the conclusion was that while not high priority we would very much like to see this issue fixed! The `--enable-valgrind` option [mentioned by @apasel422 above](https://github.com/rust-lang/rust/issues/28224#issuecomment-138725566) seems like the best fix here, and we'd just want to investigate if that has any runtime impact or affect on runtime dependencies.

**Comment by michaelwoerister** [comment]
It seems that Valgrind support in jemalloc is going away:
https://github.com/jemalloc/jemalloc/issues/369

**Comment by nathanaeljones** [comment]
I have a mixed C/Rust project targeting windows, linux, & mac. Lots of codecs are involved, and valgrind is an essential tool for verifying that integration tests aren't exposing a new bug in libpng or another dependency (this is a very common occurrence).  

Does this change mean that I should move all of my integration tests to C++? Or is a custom build of rust required?

**Comment by ssokolow** [comment]
I believe you just need to ask nightly rust to build your project with the system allocator rather than the bundled jemalloc.

**Comment by shepmaster** [comment]
Or if you produce a shared library, it will use the system allocator instead of jemalloc by default, which should allow valgrind to work.

**Comment by nathanaeljones** [comment]
@ssokolow You mean the 'official' nightly build would do this? What flags would I provide?

**Comment by shepmaster** [comment]
The [documentation covers specifying allocators](https://doc.rust-lang.org/book/custom-allocators.html).

**Comment by nathanaeljones** [comment]
Thank you! The "--enable-valgrind" context led me to misunderstand @ssokolow.  

So to clarify, there's no compilation flag that can change this; I would need to inject source code `#[feature(alloc_system)] extern crate alloc_system;` into my tests, but only as a permutation for the nightly channel?

**Comment by ssokolow** [comment]
Yes. As I understand the situation, compiling jemalloc with `--enable-valgrind` either is or was a temporary measure before support will be/was dropped entirely.

**Comment by nathanaeljones** [comment]
There is the problem that I'm running valgrind against a totally different executable, compiled by nightly instead of stable. Essentially this means I can't valgrind products of a stable rust build.

**Comment by ssokolow** [comment]
@nathanaeljones I'm no expert, but you could look into making a custom rust build with an older jemalloc and/or `--enable-valgrind` (as appropriate) while you wait for alloc_system support to be stable.

**Comment by nathanaeljones** [comment]
@ssokolow Thanks! For now it looks like nightly will be a requirement for other reasons anyway :) 

Valgrind has been crucial for my work with unsafe Rust. As far as I can tell there's no way to make the generated test harnesses Valgrind-friendly; a standalone executable is required.

**Comment by ssokolow** [comment]
As a quick update, since this is getting linked to, here's how I make the system allocator opt-in for nightly:

First, add these lines to the top of main.rs:

```rust
#![cfg_attr(feature="nightly", feature(alloc_system))]
#[cfg(feature="nightly")]
extern crate alloc_system;
```
Then, add these to Cargo.toml...
```rust
[features]
nightly = []
```
And, finally, use these commands for setup and building:
```rust
rustup toolchain add nightly-x86_64-unknown-linux-gnu
cargo +nightly build --features=nightly
```

For records purposes:
* I pieced that together for an automation "script" I haven't yet released, where I want to crunch down the size of a musl build.
* I first posted those instructions [here on reddit](https://www.reddit.com/r/rust/comments/5tgxmn/optimizing_rc_memory_usage_in_rust/ddmivgi/).
* I later cleaned up the reusable bits (so far) of the project I'd developed them in and published [this CLI project boilerplate](https://github.com/ssokolow/rust-cli-boilerplate), which incorporates that code and splits it between the `just install-rustup-deps` and `just channel=nightly build-release` commands.

**Comment by alexcrichton** [comment]
We're unlikely to really do anything to fix this, so I'm going to close this in favor of https://github.com/rust-lang/rust/issues/27389. It's highly likely that all programs will start to link to jemalloc by default once we stabilize that feature.

**Comment by fschutt** [comment]
Well, even with `alloc_system`, I still don't get any diagnostics from valgrind:

```rust
#![feature(alloc_system)]
extern crate alloc_system;

extern "C" {
  fn malloc(len: usize) -> *mut u8;
}

fn main() {
  println!("Allocated at {:?}",unsafe{malloc(100)});
}
```

```
rustc 1.25.0-nightly (45fba43b3 2018-02-10)
binary: rustc
commit-hash: 45fba43b3d5b4d1944268cf973099bfacb11bf4c
commit-date: 2018-02-10
host: x86_64-unknown-linux-gnu
release: 1.25.0-nightly
LLVM version: 6.0
```

```
==8713== Memcheck, a memory error detector
==8713== Copyright (C) 2002-2015, and GNU GPL'd, by Julian Seward et al.
==8713== Using Valgrind-3.12.0.SVN and LibVEX; rerun with -h for copyright info
==8713== Command: ./target/release/syntest
==8713==
./target/release/syntest: error while loading shared libraries: libstd-58a9e2944951d97f.so: cannot open shared object file: No such file or directory
==8713== Jump to the invalid address stated on the next line
==8713==    at 0x596: ???
==8713==    by 0x400F622: _dl_signal_error (dl-error.c:125)
==8713==    by 0x400E1D5: _dl_map_object_deps (dl-deps.c:686)
==8713==    by 0x40036D1: dl_main (rtld.c:1729)
==8713==    by 0x4017AFD: _dl_sysdep_start (dl-sysdep.c:249)
==8713==    by 0x4001AE7: _dl_start_final (rtld.c:401)
==8713==    by 0x4001AE7: _dl_start (rtld.c:507)
==8713==    by 0x4000C27: ??? (in /lib/x86_64-linux-gnu/ld-2.24.so)
==8713==  Address 0x596 is not stack'd, malloc'd or (recently) free'd
==8713==
==8713==
==8713== Process terminating with default action of signal 11 (SIGSEGV)
==8713==  Bad permissions for mapped region at address 0x596
==8713==    at 0x596: ???
==8713==    by 0x400F622: _dl_signal_error (dl-error.c:125)
==8713==    by 0x400E1D5: _dl_map_object_deps (dl-deps.c:686)
==8713==    by 0x40036D1: dl_main (rtld.c:1729)
==8713==    by 0x4017AFD: _dl_sysdep_start (dl-sysdep.c:249)
==8713==    by 0x4001AE7: _dl_start_final (rtld.c:401)
==8713==    by 0x4001AE7: _dl_start (rtld.c:507)
==8713==    by 0x4000C27: ??? (in /lib/x86_64-linux-gnu/ld-2.24.so)
==8713==
==8713== HEAP SUMMARY:
==8713==     in use at exit: 0 bytes in 0 blocks
==8713==   total heap usage: 0 allocs, 0 frees, 0 bytes allocated
==8713==
==8713== All heap blocks were freed -- no leaks are possible
==8713==
==8713== For counts of detected and suppressed errors, rerun with: -v
==8713== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
```

Is there a workaround?
